plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Contexts.hs

```diff
@@ -0,0 +1,282 @@
+    (
+    -- * Pending transactions and related types
+      TxInfo(..)
+    , ScriptContext(..)
+    , ScriptPurpose(..)
+    , TxId (..)
+    , TxOut(..)
+    , TxOutRef(..)
+    , TxInInfo(..)
+    , findOwnInput
+    , findDatum
+    , findDatumHash
+    , findTxInByTxOutRef
+    , findContinuingOutputs
+    , getContinuingOutputs
+    -- * Validator functions
+    , pubKeyOutputsAt
+    , valuePaidTo
+    , spendsOutput
+    , txSignedBy
+    , valueSpent
+    , valueProduced
+    , ownCurrencySymbol
+    ) where
+
+import GHC.Generics (Generic)
+import PlutusTx
+import PlutusTx.Prelude
+import Prettyprinter
+import Prettyprinter.Extras
+
+import PlutusLedgerApi.V1.Address (Address (..))
+import PlutusLedgerApi.V1.Credential (Credential (..), StakingCredential)
+import PlutusLedgerApi.V1.Crypto (PubKeyHash (..))
+import PlutusLedgerApi.V1.Data.Tx (TxId (..), TxOut (..), TxOutRef (..))
+import PlutusLedgerApi.V1.Data.Value (CurrencySymbol (..), Value)
+import PlutusLedgerApi.V1.DCert (DCert (..))
+import PlutusLedgerApi.V1.Scripts
+import PlutusLedgerApi.V1.Time (POSIXTimeRange)
+import Prelude qualified as Haskell
+
+{- Note [Script types in pending transactions]
+To validate a transaction, we have to evaluate the validation script of each of
+the transaction's inputs. The validation script sees the data of the
+transaction output it validates, and the redeemer of the transaction input of
+the transaction that consumes it.
+In addition, the validation script also needs information on the transaction as
+a whole (not just the output-input pair it is concerned with). This information
+is provided by the `TxInfo` type. A `TxInfo` contains the hashes of
+redeemer and data scripts of all of its inputs and outputs.
+-}
+
+-- | An input of a pending transaction.
+data TxInInfo = TxInInfo
+    { txInInfoOutRef   :: TxOutRef
+    , txInInfoResolved :: TxOut
+    } deriving stock (Generic, Haskell.Show, Haskell.Eq)
+
+instance Eq TxInInfo where
+    TxInInfo ref res == TxInInfo ref' res' = ref == ref' && res == res'
+
+instance Pretty TxInInfo where
+    pretty TxInInfo{txInInfoOutRef, txInInfoResolved} =
+        pretty txInInfoOutRef <+> "->" <+> pretty txInInfoResolved
+
+-- | Purpose of the script that is currently running
+data ScriptPurpose
+    = Minting CurrencySymbol
+    | Spending TxOutRef
+    | Rewarding StakingCredential
+    | Certifying DCert
+    deriving stock (Generic, Haskell.Show, Haskell.Eq)
+    deriving Pretty via (PrettyShow ScriptPurpose)
+
+instance Eq ScriptPurpose where
+    {-# INLINABLE (==) #-}
+    Minting cs == Minting cs'           = cs == cs'
+    Spending ref == Spending ref'       = ref == ref'
+    Rewarding sc == Rewarding sc'       = sc == sc'
+    Certifying cert == Certifying cert' = cert == cert'
+    _ == _                              = False
+
+-- | A pending transaction. This is the view as seen by validator scripts, so some details are stripped out.
+data TxInfo = TxInfo
```

Comment:
This looks the same as the existing `TxInfo`, so why redefine it?

Reply:
It uses different Map and Value types, so it's not the same TxInfo.